client/src/utils/CountdownTimer.tsx

```diff
@@ -0,0 +1,54 @@
+import React, { useState, useEffect } from 'react'
+
+interface TimeLeft {
+  days: number
+  hours: number
+  minutes: number
+  seconds: number
+}
+
+export function CountdownTimer(props: { deadline: Date }) {
+  function calculateTimeLeft(): TimeLeft {
+    const difference: number = props.deadline.getTime() - Date.now()
+
+    if (difference > 0) {
+      return {
+        days: Math.floor(difference / (1000 * 60 * 60 * 24)),
+        hours: Math.floor((difference / (1000 * 60 * 60)) % 24),
+        minutes: Math.floor((difference / 1000 / 60) % 60),
+        seconds: Math.floor((difference / 1000) % 60),
+      }
+    }
+    return {
+      days: 0,
+      hours: 0,
+      minutes: 0,
+      seconds: 0,
+    }
+  }
+
+  const [timeLeft, setTimeLeft] = useState<TimeLeft>(calculateTimeLeft())
+
+  useEffect(() => {
+    setTimeout(() => {
+      setTimeLeft(calculateTimeLeft())
+    }, 1000)
+  })
```

Comment:
We need to remove the timeout when the component unmounts, otherwise we might try to call the `setTimeLeft` on a component that was already unmounted. To do that [the function we pass into `useEffect` should return the function for the cleanup](https://reactjs.org/docs/hooks-reference.html#cleaning-up-an-effect).

This should do the trick
```ts
  useEffect(() => {
    const timeout = setTimeout(() => {
      setTimeLeft(calculateTimeLeft())
    }, 1000)
    return () => clearTimeout(timeout)
  })
```

Reply:
👍